client/allocrunner/task_coordinator.go

```diff
@@ -0,0 +1,386 @@
+	taskLifecycleMain = "main"
+
+	// taskLifecyclePoststart are tasks with the "poststart" hook.
+	taskLifecyclePoststart = "poststart"
+
+	// taskLifecyclePoststop are tasks with the "poststop" hook.
+	taskLifecyclePoststop = "poststop"
+)
+
+// taskCoordinator controls when tasks with a given lifecycle configuration are
+// allowed to run.
+//
+// It behaves like a finite state machine where each state transition blocks or
+// allows some task lifecycle type to run.
+type taskCoordinator struct {
+	logger hclog.Logger
+
+	// tasksByLifecycle is an index used to group and quickly access tasks by
+	// their lifecycle configuration.
+	tasksByLifecycle map[taskLifecycle][]string
+
+	// currentState is the current state of the FSM. It must only be accessed
+	// while holding the lock.
+	currentState     taskCoordinatorState
+	currentStateLock sync.RWMutex
+
+	// controllers store a controller for each task lifecycle.
+	controllers map[taskLifecycle]*taskCoordinatorController
+}
+
+// newTaskCoordinator returns a new taskCoordinator in the "init" state and
+// with all tasks blocked.
+func newTaskCoordinator(logger hclog.Logger, tasks []*structs.Task, shutdownCh <-chan struct{}) *taskCoordinator {
+	c := &taskCoordinator{
+		logger:           logger.Named("task_coordinator"),
+		tasksByLifecycle: indexTasksByLifecycle(tasks),
+		controllers:      make(map[taskLifecycle]*taskCoordinatorController),
+	}
+
+	for lifecycle := range c.tasksByLifecycle {
+		ctrl := newTaskCoordinatorController()
+		go ctrl.run(shutdownCh)
+
+		c.controllers[lifecycle] = ctrl
+	}
+
+	c.enterStateLocked(taskCoordinatorStateInit)
+	return c
+}
+
+// restore is used to set the coordinatator FSM to the correct state when an
+// alloc is restored. Must be called before the allocrunner is running.
+func (c *taskCoordinator) restore(states map[string]*structs.TaskState) {
+	// Skip the "init" state when restoring since the tasks were likelly
+	// already running.
+	c.currentStateLock.Lock()
+	c.enterStateLocked(taskCoordinatorStatePrestart)
+	c.currentStateLock.Unlock()
+
+	c.taskStateUpdated(states)
+}
+
+// startConditionForTask returns a channel that a task runner can use to wait
+// until it is allowed to run.
+func (c *taskCoordinator) startConditionForTask(task *structs.Task) <-chan struct{} {
+	var lifecycle taskLifecycle
+
+	if task.IsPrestart() {
+		if task.Lifecycle.Sidecar {
+			lifecycle = taskLifecyclePrestartSidecar
+		} else {
+			lifecycle = taskLifecyclePrestartEphemeral
+		}
+	} else if task.IsPoststart() {
+		lifecycle = taskLifecyclePoststart
+	} else if task.IsPoststop() {
+		lifecycle = taskLifecyclePoststop
+	} else {
+		// Assume task is "main" by default.
+		lifecycle = taskLifecycleMain
+	}
+
+	return c.controllers[lifecycle].waitCh()
+}
+
+// taskStateUpdated notifies that a task state has changed. This may cause the
+// taskCoordinator FSM to progresses to another state.
+func (c *taskCoordinator) taskStateUpdated(states map[string]*structs.TaskState) {
```

Comment:
A note to any other reviewers: these `*structs.TaskState` pointers are to copies of the task state created in allocrunner